src/commands/index.rs

```diff
@@ -0,0 +1,163 @@
+    use std::collections::HashMap;
+
+    type SchemaMap = HashMap<String, Schema>;
+
+    /// We keep a static map, since it needs to be accessed from multiple Redis module commands
+    /// and there is no other obvious way to implement this (such as a "user data" pointer maintained
+    /// for us by Redis).
+    ///
+    /// The init function should be called only once.
+    /// Unwrapping the Option is thus safe afterwards.
+    ///
+    /// Since we have only one thread at the moment, getting a (mutable) reference to the map
+    /// is also safe. If we add threads, we can simply wrap the map: `Option<Mutex<SchemaMap>`.
+    ///
+    static mut SCHEMA_MAP: Option<SchemaMap> = None;
+
+    pub fn init() {
+        let map = HashMap::new();
+        unsafe {
+            SCHEMA_MAP = Some(map);
+        }
+    }
+
+    pub fn as_ref() -> &'static SchemaMap {
+        unsafe { SCHEMA_MAP.as_ref() }.unwrap()
+    }
+
+    pub fn as_mut() -> &'static mut SchemaMap {
+        unsafe { SCHEMA_MAP.as_mut() }.unwrap()
+    }
+}
+
+///////////////////////////
+
+fn add_field<I>(index_name: &str, field_name: &str, mut args: I) -> RedisResult
+where
+    I: Iterator<Item = String>,
+{
+    let path = args.next_string()?;
+
+    let map = schema_map::as_mut();
+
+    if !map.contains_key(index_name) {
+        let schema = Schema::new(&index_name);
+        map.insert(index_name.to_owned(), schema);
+    }
+
+    let schema = map.get_mut(index_name).unwrap();
+
+    if schema.fields.contains_key(field_name) {
+        Err("Field already exists".into())
+    } else {
+        schema.index.create_field(field_name);
+        schema.fields.insert(field_name.to_string(), path);
+        REDIS_OK
+    }
+}
+
+pub fn add_document(key: &str, index_name: &str, doc: &RedisJSON) -> RedisResult {
+    // TODO: Index the document with RediSearch:
+    // 1. Determine the index to use (how?)
+    // 2. Get the fields from the index along with their associated paths
+    // 3. Build a RS Document and populate the fields from our doc
+    // 4. Add the Document to the index
+
+    let map = schema_map::as_ref();
+
+    map.get(index_name)
+        .ok_or("ERR no such index".into())
+        .and_then(|schema| {
+            let rsdoc = create_document(key, schema, doc)?;
+            schema.index.add_document(&rsdoc)?;
+            REDIS_OK
+        })
+}
+
+fn create_document(key: &str, schema: &Schema, doc: &RedisJSON) -> Result<Document, Error> {
+    let fields = &schema.fields;
+
+    let score = 1.0;
+    let rsdoc = Document::create(key, score);
+
+    for (field_name, path) in fields {
+        let value = doc.get_doc(&path)?;
+
+        match value {
+            Value::String(v) => rsdoc.add_field(field_name, &v, FieldType::FULLTEXT),
+            Value::Number(v) => rsdoc.add_field(field_name, &v.to_string(), FieldType::NUMERIC),
+            Value::Bool(v) => rsdoc.add_field(field_name, &v.to_string(), FieldType::TAG),
+            _ => {}
```

Comment:
@gavrie should we also handle Objects and Array?

Reply:
Good question; I discussed this with Meir a bit -- but what would the semantics be? Let's discuss f2f.